lib/ocpp/v201/messages/Authorize.cpp

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2023 Pionix GmbH and Contributors to EVerest
+
+#include <ostream>
+#include <string>
+
+#include <boost/optional/optional.hpp>
+#include <nlohmann/json.hpp>
+
+#include <ocpp/v201/enums.hpp>
+#include <ocpp/v201/messages/Authorize.hpp>
+#include <ocpp/v201/ocpp_types.hpp>
+
+using json = nlohmann::json;
+
+namespace ocpp {
+namespace v201 {
+
+std::string AuthorizeRequest::get_type() const {
+    return "Authorize";
+}
+
+void to_json(json& j, const AuthorizeRequest& k) {
+    // the required parts of the message
+    j = json{
+        {"idToken", k.idToken},
+    };
+    // the optional parts of the message
+    if (k.customData) {
+        j["customData"] = k.customData.value();
+    }
+    if (k.certificate) {
+        j["certificate"] = k.certificate.value();
+    }
+    if (k.iso15118CertificateHashData) {
+        j["iso15118CertificateHashData"] = json::array();
+        for (auto val : k.iso15118CertificateHashData.value()) {
+            j["iso15118CertificateHashData"].push_back(val);
+        }
+    }
+}
+
+void from_json(const json& j, AuthorizeRequest& k) {
+    // the required parts of the message
+    k.idToken = j.at("idToken");
+
+    // the optional parts of the message
+    if (j.contains("customData")) {
+        k.customData.emplace(j.at("customData"));
+    }
+    if (j.contains("certificate")) {
+        k.certificate.emplace(j.at("certificate"));
+    }
+    if (j.contains("iso15118CertificateHashData")) {
+        json arr = j.at("iso15118CertificateHashData");
+        std::vector<OCSPRequestData> vec;
+        for (auto val : arr) {
+            vec.push_back(val);
+        }
+        k.iso15118CertificateHashData.emplace(vec);
+    }
+}
+
+/// \brief Writes the string representation of the given AuthorizeRequest \p k to the given output stream \p os
+/// \returns an output stream with the AuthorizeRequest written to
+std::ostream& operator<<(std::ostream& os, const AuthorizeRequest& k) {
+    os << json(k).dump(4);
```

Comment:
we should probably have a constant for this indentation of 4 characters defined somewhere